libraries/freertos_plus/standard/https/src/iot_https_client.c

```diff
@@ -0,0 +1,1829 @@
+        return IOT_HTTPS_PARSING_ERROR;
+    }
+
+    return IOT_HTTPS_OK;
+}
+
+/*-----------------------------------------------------------*/
+
+static IotHttpsReturnCode_t _receiveHttpsMessage( _httpsConnection_t* _httpsConnection, 
+    http_parser* pParser,
+    IotHttpsResponseParserState_t *pCurrentParserState,
+    IotHttpsResponseParserState_t finalParserState, 
+    uint8_t** pBuf,
+    uint8_t** pBufCur,
+    uint8_t** pBufEnd,
+    IotHttpsReturnCode_t *pNetworkStatus)
+{    
+    IotHttpsReturnCode_t status = IOT_HTTPS_OK;
+
+    while( (*pCurrentParserState < finalParserState) && (*pBufEnd - *pBufCur > 0) ) 
+    {
+        *pNetworkStatus = _networkRecv( _httpsConnection,
+            *pBufCur, 
+            *pBufEnd - *pBufCur);
+
+        status = _parseHttpsMessage(pParser, *pBufCur, *pBufEnd - *pBufCur);
+        if(status != IOT_HTTPS_OK)
+        {
+            IotLogError("Failed to parse the message buffer with error: %d", pParser->http_errno);
+            return status;
+        }
+
+        /* The _httResponse->pHeadersCur pointer is updated in the http_parser callbacks. */
+        IotLogDebug( "There is %d of space left in the buffer.", *pBufEnd - *pBufCur );
+
+        /* We cannot anticipate the unique network error received when the server closes the connection. 
+           We simply exit the loop if there nothing else to receive. We do not return the network status because 
+           the error may just be a server closed the connection, but there may still have been data in the buffer
+           that we parsed. */
+        if( *pNetworkStatus != IOT_HTTPS_OK )
+        {
+            IotLogError( "Error receiving the HTTPS response headers. Error code: %d", status );
+            break;
+        }
+
+    }
+
+    /* If we did not reach the end of the headers with the current header buffer size then log that info. */
+    if( *pCurrentParserState < finalParserState)
+    {
+        IotLogDebug("There are still more data on the network. It could not fit into buffer at \
+            location 0x%x with length %d.",
+            *pBuf,
+            *pBufEnd - *pBuf);
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static IotHttpsReturnCode_t _receiveHttpsHeaders( _httpsConnection_t* _httpsConnection, _httpsResponse_t* _httpsResponse, IotHttpsReturnCode_t *pNetworkStatus)
+{
+    IotHttpsReturnCode_t status = IOT_HTTPS_OK;
+
+    /* Initialize the HTTPS parser for a new execution. */
+    http_parser_init(&(_httpsResponse->httpParser), HTTP_RESPONSE);
+    _httpsResponse->httpParser.data = (void *)(_httpsResponse);
+    _httpsResponse->parserState = PARSER_STATE_NONE;
+    _httpsResponse->bufferProcessingState = PROCESSING_STATE_FILLING_HEADER_BUFFER;
+
+    status = _receiveHttpsMessage(_httpsConnection,
+        &(_httpsResponse->httpParser),
+        &(_httpsResponse->parserState),
+        PARSER_STA
```

Comment:
Is ```respHandle->pHeadersCur``` pointing to the last header? In this case, is it possible that the parser will never look for the specified header in the last header in this buffer?

Reply:
Yes that is pointing to the byte after last header field or value byte processed. it is possible that the parser will never find the last header. That is fine. When we go to read the body, those headers will be filled in the body buffer and the pointers for bodyCur will not be moved until body is actually read from the network.